corehq/apps/registration/utils.py

```diff
@@ -313,3 +321,89 @@ def send_mobile_experience_reminder(recipient, full_name, additional_email_conte
         logging.warning(
             "Can't send email, but the message was:\n%s" % message_plaintext)
         raise
+
+
+@run_only_when(not settings.ENTERPRISE_MODE and not settings.UNIT_TESTING)
+@silence_and_report_error("Exception raised in the check username availability rate limiter",
+                          'commcare.registration.check_username_rate_limiter_errors')
+def rate_limit_check_username_availability():
+    """
+    This holds attempts per IP OR per session below limits given by
+    check_username_rate_limiter_by_ip and check_username_rate_limiter_by_session, respectively.
+
+    Requests without both an IP and session ID are rejected.
+    """
+    _status_session_rate_limited = 'session_rate_limited'
+    _status_ip_rate_limited = 'ip_rate_limited'
+    _status_bad_request = 'bad_request'
+    _status_accepted = 'accepted'
+
+    def get_session_and_ip():
+        request = get_request()
+        if request and request.session:
+            return request.session.session_key, get_ip(request)
+        else:
+            return None
+
+    def _check_for_exceeded_rate_limits(session, ip):
+        session_window = (
+            check_username_rate_limiter_per_session
+            .get_window_of_first_exceeded_limit('session:{}'.format(session))
+        )
+        ip_window = check_username_rate_limiter_per_ip.get_window_of_first_exceeded_limit('ip:{}'.format(ip))
+
+        # ensure that no rate limit windows have been exceeded
+        # order of priority (session, ip)
+        if session_window is not None:
+            return _status_session_rate_limited, session_window
+        elif ip_window is not None:
+            return _status_ip_rate_limited, ip_window
+        else:
+            return _status_accepted, None
+
+    session_id, ip_address = get_session_and_ip()
+
+    if ip_address and session_id:
```

Comment:
Given that these are bad-faith actors, is it possible to continually create new session ids? Is this a concern?

Reply:
Not as far as I've found. If given a bad session ID, the request is rejected on the backend and the session ID cookie gets removed.